File: benchmarks/workloads/ecs/ecs.c
// ECS update: structure-of-arrays storage, bitmask queries, 1M entities.
// Timed region: 1000 ticks of movement, damage, and cleanup systems.
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define ENTITIES 1000000
#define TICKS 1000
#define HAS_POS 1
#define HAS_VEL 2
#define HAS_HP 4
#define HAS_DMG 8

typedef struct { float x, y; } Position;
typedef struct { float x, y; } Velocity;
typedef struct { float hp; } Health;
typedef struct { float dps; } Damage;

typedef struct {
    uint8_t *mask;
    Position *pos;
    Velocity *vel;
    Health *hp;
    Damage *dmg;
    int count;
} World;

static double now_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000.0 + ts.tv_nsec / 1000000.0;
}

static World world_new(void) {
    World w;
    w.mask = calloc(ENTITIES, sizeof(uint8_t));
    w.pos = calloc(ENTITIES, sizeof(Position));
    w.vel = calloc(ENTITIES, sizeof(Velocity));
    w.hp = calloc(ENTITIES, sizeof(Health));
    w.dmg = calloc(ENTITIES, sizeof(Damage));
    w.count = 0;
    return w;
}

static void system_movement(World *w, float dt) {
    const uint8_t need = HAS_POS | HAS_VEL;
    for (int i = 0; i < w->count; i++) {
        if ((w->mask[i] & need) == need) {
            w->pos[i].x += w->vel[i].x * dt;
            w->pos[i].y += w->vel[i].y * dt;
        }
    }
}

static void system_damage(World *w, float dt) {
    const uint8_t need = HAS_HP | HAS_DMG;
    for (int i = 0; i < w->count; i++) {
        if ((w->mask[i] & need) == need) w->hp[i].hp -= w->dmg[i].dps * dt;
    }
}

static void system_cleanup(World *w) {
    for (int i = 0; i < w->count; i++) {
        if ((w->mask[i] & HAS_HP) != 0 && w->hp[i].hp <= 0.0f) w->mask[i] = 0;
    }
}

static int count_alive(const World *w) {
    int n = 0;
    for (int i = 0; i < w->count; i++) if (w->mask[i] != 0) n++;
    return n;
}

int main(void) {
    World w = world_new();
    // 70% move, 50% have health, 30% take damage.
    for (uint32_t i = 0; i < ENTITIES; i++) {
        int id = w.count++;
        float fi = (float)i;
        if (i % 10 < 7) {
            w.mask[id] |= HAS_POS; w.pos[id] = (Position){ fi * 0.1f, fi * 0.2f };
            w.mask[id] |= HAS_VEL; w.vel[id] = (Velocity){ ((float)(i % 100)) * 0.01f, ((float)((i + 50) % 100)) * 0.01f };
        }
        if (i % 10 < 5) { w.mask[id] |= HAS_HP; w.hp[id] = (Health){ 100.0f }; }
        if (i % 10 < 3) { w.mask[id] |= HAS_DMG; w.dmg[id] = (Damage){ 0.5f + ((float)(i % 10)) * 0.1f }; }
    }
    int alive_before = count_alive(&w);

    double start = now_ms();
    float dt = 1.0f / 60.0f;
    for (int t = 0; t < TICKS; t++) {
        system_movement(&w, dt);
        system_damage(&w, dt);
        system_cleanup(&w);
    }
    double elapsed_ms = now_ms() - start;

    double sum = 0.0;
    for (int i = 0; i < w.count; i++) if (w.mask[i] & HAS_POS) sum += (double)w.pos[i].x;
    printf("entities %d alive_before %d alive_after %d\n", w.count, alive_before, count_alive(&w));
    printf("elapsed_ms %.3f\n", elapsed_ms);
    printf("checksum %.2f\n", sum);
    free(w.mask); free(w.pos); free(w.vel); free(w.hp); free(w.dmg);
    return 0;
}

File: benchmarks/workloads/ecs/ecs.go
// ECS update: structure-of-arrays storage, bitmask queries, 1M entities.
// Timed region: 1000 ticks of movement, damage, and cleanup systems.
package main

import (
	"fmt"
	"time"
)

const (
	entities = 1000000
	ticks    = 1000
	hasPos   = 1
	hasVel   = 2
	hasHP    = 4
	hasDmg   = 8
)

type Position struct{ x, y float32 }
type Velocity struct{ x, y float32 }
type Health struct{ hp float32 }
type Damage struct{ dps float32 }

type World struct {
	mask  []uint8
	pos   []Position
	vel   []Velocity
	hp    []Health
	dmg   []Damage
	count int
}

func newWorld() *World {
	return &World{
		mask: make([]uint8, entities),
		pos:  make([]Position, entities),
		vel:  make([]Velocity, entities),
		hp:   make([]Health, entities),
		dmg:  make([]Damage, entities),
	}
}

func (w *World) create() int {
	id := w.count
	w.count++
	return id
}

// Explicit float32 conversions on the products keep Go from fusing them into
// FMAs on arm64, so every language computes the same rounded f32 values.
func (w *World) systemMovement(dt float32) {
	const need = hasPos | hasVel
	for i := 0; i < w.count; i++ {
		if w.mask[i]&need == need {
			w.pos[i].x += float32(w.vel[i].x * dt)
			w.pos[i].y += float32(w.vel[i].y * dt)
		}
	}
}

func (w *World) systemDamage(dt float32) {
	const need = hasHP | hasDmg
	for i := 0; i < w.count; i++ {
		if w.mask[i]&need == need {
			w.hp[i].hp -= float32(w.dmg[i].dps * dt)
		}
	}
}

func (w *World) systemCleanup() {
	for i := 0; i < w.count; i++ {
		if w.mask[i]&hasHP != 0 && w.hp[i].hp <= 0 {
			w.mask[i] = 0
		}
	}
}

func (w *World) countAlive() int {
	n := 0
	for i := 0; i < w.count; i++ {
		if w.mask[i] != 0 {
			n++
		}
	}
	return n
}

func main() {
	w := newWorld()
	// 70% move, 50% have health, 30% take damage.
	for i := uint32(0); i < entities; i++ {
		id := w.create()
		fi := float32(i)
		if i%10 < 7 {
			w.mask[id] |= hasPos
			w.pos[id] = Position{float32(fi * 0.1), float32(fi * 0.2)}
			w.mask[id] |= hasVel
			w.vel[id] = Velocity{float32(float32(i%100) * 0.01), float32(float32((i+50)%100) * 0.01)}
		}
		if i%10 < 5 {
			w.mask[id] |= hasHP
			w.hp[id] = Health{100}
		}
		if i%10 < 3 {
			w.mask[id] |= hasDmg
			w.dmg[id] = Damage{0.5 + float32(float32(i%10)*0.1)}
		}
	}
	aliveBefore := w.countAlive()

	start := time.Now()
	dt := float32(1.0 / 60.0)
	for t := 0; t < ticks; t++ {
		w.systemMovement(dt)
		w.systemDamage(dt)
		w.systemCleanup()
	}
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6

	sum := 0.0
	for i := 0; i < w.count; i++ {
		if w.mask[i]&hasPos != 0 {
			sum += float64(w.pos[i].x)
		}
	}
	fmt.Printf("entities %d alive_before %d alive_after %d\n", w.count, aliveBefore, w.countAlive())
	fmt.Printf("elapsed_ms %.3f\n", elapsedMs)
	fmt.Printf("checksum %.2f\n", sum)
}
